tests/unit/test_instance_connection_manager.py

```diff
@@ -53,13 +55,99 @@ def test_InstanceConnectionManager_init(async_loop: asyncio.AbstractEventLoop) -
 
 
 @pytest.mark.asyncio
-async def test_InstanceConnectionManager_perform_refresh(
-    icm: InstanceConnectionManager, async_loop: asyncio.AbstractEventLoop
+async def test_perform_refresh_replaces_result(icm: InstanceConnectionManager) -> None:
+    """
+    Test to check whether _perform_refresh replaces a valid result with another valid result
+    """
+
+    class MockMetadata:
+        def __init__(self, expiration: datetime.datetime) -> None:
+            self.expiration = expiration
+
+    async def _get_metadata_success(*args: Any, **kwargs: Any) -> MockMetadata:
+        return MockMetadata(datetime.datetime.now() + datetime.timedelta(minutes=10))
+
+    # stub _get_instance_data to return a "valid" MockMetadata object
+    setattr(icm, "_get_instance_data", _get_metadata_success)
+    new_task = await icm._perform_refresh()
+
+    # allow enough time for the task to complete
+    await asyncio.sleep(10)
```

Comment:
I get a "Task attached to a different loop" error (probably because tests are running from a different OS thread?). I agree that the sleep is pretty hacky and gross. Is there a better way to get rid of that error? I tried using `asyncio.set_event_loop` to set it to the loop that the icm is using, but it didn't work and I don't really understand why it didn't

Reply:
Oh interesting. 

Can you try using something like `asyncio.run_coroutine_threadsafe(() -> await new_task, icm._loop).result(timeout=XX)`?

Reply:
`run_coroutine_threadsafe` fixed it! I actually had to use it to call _perform_refresh because the issue was that I was calling it from the test thread.